torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp

```diff
@@ -596,6 +596,87 @@ class TORCH_API ProcessGroupNCCL : public Backend {
     std::string traceKeyEnd_;
   };
 
+  // Class that runs as a separate thread aside from watchdog
+  // thread because we need to check the heartbeat from watchdog thread
+  // so that when we get stuck in some NCCL/CUDA calls,
+  // we can dump the debugging information and abort the process.
+  class HeartbeatMonitor {
+   public:
+    HeartbeatMonitor(ProcessGroupNCCL* pg);
+    virtual ~HeartbeatMonitor() = default;
+
+    // Start the heartbeat monitor thread.
+    void start();
+
+    // Stop the heartbeat monitor thread.
+    void wait();
+
+    // Run the actual loop to check watchdog heartbeat.
+    virtual void runLoop();
+
+    // notify the heartbeat monitor thread to stop.
+    void stop();
+
+    // print out ENV settings for the heartbeat monitor thread.
+    void printLogMsg();
```

Comment:
Move this into constructor thus no need for an API?

Reply:
Ahh ok..